PseudoGantry/lib/PseudoGantry/PseudoAxis.cpp

```diff
@@ -0,0 +1,121 @@
+#include "PseudoAxis.h"
+
+void set_up_encoder(PseudoEncoder *encoder, void (*isr_motor_pulse)(void))
+{
+    pinMode(encoder->motor_pulse_pin, INPUT_PULLUP);
+    pinMode(encoder->channel_a_out, OUTPUT);
+    digitalWrite(encoder->channel_a_out, LOW);
+    attachInterrupt(digitalPinToInterrupt(encoder->motor_pulse_pin), isr_motor_pulse, CHANGE);
+}
+
+bool toggle_encoder_output(PseudoAxis *pseudo_axis)
+{
+    pseudo_axis->skip_counter++;
+    // skip steps in the beginning
+    if (pseudo_axis->skip_counter <= pseudo_axis->changes_to_skip) return false;
+    // else start toggling until the next set of STEPS
+    else if ((pseudo_axis->skip_counter <= (2 * pseudo_axis->steps_for_ratio)) || (pseudo_axis->changes_to_skip == 0))
+    {   
+        bool motor_pin_status = digitalRead(pseudo_axis->encoder.motor_pulse_pin);
+        // Serial.print("motor pin:    "); Serial.println(motor_pin_status);
+        digitalWrite((pseudo_axis->encoder).channel_a_out, motor_pin_status ^ HIGH);
+        // check if encoder count needs to be incremented
+        // if HIGH it means triggered on rising
+        if (motor_pin_status) return true;
+        else return false;
+    }
+    else
+    {
+        // this is assigned 1 and not 0 because at this point our step_count_x is ((2 * STEPS) + 1)
+        // which means we are onto the next set of 8, and 17 would be the beginnig of that set
+        pseudo_axis->skip_counter = 1;
```

Comment:
If `steps_for_ratio` and `counts_for_ratio` are the same, then `changes_to_skip` will be `0`.
In that case we will never reach this statement (since the `else if` above will always evaluate to true). This means the `skip_counter` will count up until it wraps around to negative numbers (since it's declared as an `int`). Once it reaches negative numbers, the first `if (pseudo_axis->skip_counter <= pseudo_axis->changes_to_skip)` will start evaluating to `true` and we'll stop toggling the encoder output pin.
The problem wouldn't appear until we reach 2 billion counts, but if it's not too complicated to handle correctly I think it's worth doing.

Reply:
damn, that is a really good catch!!! I just changed it to if `(pseudo_axis->changes_to_skip != 0) pseudo_axis->change_counter++;`

Reply:
I believe that almost fixes it, but `change_counter` starts set to 0, which means if `changes_to_skip` is 0, it'll never be incremented.
Then `if (pseudo_axis->change_counter <= pseudo_axis->changes_to_skip)` will always evaluate to true (since both values are 0).

You could fix it by setting `change_counter` to 1 in `reset_pseudo_axis` only if `changes_to_skip` is 0 although that's kinda hacky.

Better approach might be to replace:
```c
if (pseudo_axis->change_counter <= pseudo_axis->changes_to_skip)
```
with
```c
if ((pseudo_axis->change_counter <= pseudo_axis->changes_to_skip) && (pseudo_axis->changes_to_skip != 0))
```

Reply:
Good idea! Fixed